arbos/l1pricing/l1pricing.go

```diff
@@ -7,239 +7,365 @@ import (
-func (ps *L1PricingState) L1BaseFeeEstimateInertia() (uint64, error) {
-	return ps.l1BaseFeeEstimateInertia.Get()
+func (ps *L1PricingState) FundsDueForRewards() (*big.Int, error) {
+	return ps.fundsDueForRewards.Get()
 }
 
-// Set how slowly ArbOS updates its estimate of the L1 basefee
-func (ps *L1PricingState) SetL1BaseFeeEstimateInertia(inertia uint64) error {
-	return ps.l1BaseFeeEstimateInertia.Set(inertia)
+func (ps *L1PricingState) SetFundsDueForRewards(amt *big.Int) error {
+	return ps.fundsDueForRewards.Set(amt)
 }
 
-func (ps *L1PricingState) userSpecifiedAggregatorsForAddress(sender common.Address) *addressSet.AddressSet {
-	return addressSet.OpenAddressSet(ps.userSpecifiedAggregators.OpenSubStorage(sender.Bytes()))
+func (ps *L1PricingState) UnitsSinceUpdate() (uint64, error) {
+	return ps.unitsSinceUpdate.Get()
 }
 
-// Get sender's user-specified aggregator, or nil if there is none. This does NOT fall back to the default aggregator
-//     if there is no user-specified aggregator. If that is what you want, call ReimbursableAggregatorForSender instead.
-func (ps *L1PricingState) UserSpecifiedAggregator(sender common.Address) (*common.Address, error) {
-	return ps.userSpecifiedAggregatorsForAddress(sender).GetAnyMember()
+func (ps *L1PricingState) SetUnitsSinceUpdate(units uint64) error {
+	return ps.unitsSinceUpdate.Set(units)
 }
 
-func (ps *L1PricingState) SetUserSpecifiedAggregator(sender common.Address, maybeAggregator *common.Address) error {
-	paSet := ps.userSpecifiedAggregatorsForAddress(sender)
-	if err := paSet.Clear(); err != nil {
+func (ps *L1PricingState) AddToUnitsSinceUpdate(units uint64) error {
+	oldUnits, err := ps.unitsSinceUpdate.Get()
+	if err != nil {
 		return err
 	}
-	if maybeAggregator == nil {
-		return nil
-	}
-	return paSet.Add(*maybeAggregator)
+	return ps.unitsSinceUpdate.Set(oldUnits + units)
+}
+
+func (ps *L1PricingState) PricePerUnit() (*big.Int, error) {
+	return ps.pricePerUnit.Get()
+}
+
+func (ps *L1PricingState) SetPricePerUnit(price *big.Int) error {
+	return ps.pricePerUnit.Set(price)
 }
 
-func (ps *L1PricingState) RefusesDefaultAggregator(addr common.Address) (bool, error) {
-	val, err := ps.refuseDefaultAggregator.Get(common.BytesToHash(addr.Bytes()))
+// Update the pricing model based on a payment by a batch poster
+func (ps *L1PricingState) UpdateForBatchPosterSpending(statedb vm.StateDB, evm *vm.EVM, updateTime uint64, currentTime uint64, batchPoster common.Address, weiSpent *big.Int) error {
+	batchPosterTable := ps.BatchPosterTable()
+	posterState, err := batchPosterTable.OpenPoster(batchPoster)
 	if err != nil {
-		return false, err
+		return err
 	}
-	return val != (common.Hash{}), nil
-}
 
-func (ps *L1PricingState) SetRefusesDefaultAggregator(addr common.Address, refuses bool) error {
-	val := uint64(0)
-	if refuses {
-		val = 1
+	// compute previous shortfall
+	totalFundsDue, err := batchPosterTable.TotalFundsDue()
```

Comment:
We do want to use the total funds due here, because there is a single charging mechanism that is supposed to reimburse all posters. Basically we want to know whether we have over or under collected. across all posters.

Reply:
Makes sense, thanks